ChatApp/View/TypingAreaView.swift

```diff
@@ -0,0 +1,146 @@
+        
+        NSLayoutConstraint.activate([
+            messageTextView.topAnchor.constraint(equalTo: topAnchor, constant: 19),
+            messageTextView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -22),
+            messageTextView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 22),
+            messageTextView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -19),
+        ])
+    }
+    
+    private func setUpMessageTextViewPlaceholder() {
+        messageTextView.addSubview(placeholderLabel)
+        
+        placeholderLabel.text = "დაწერე შეტყობინება..."
+        placeholderLabel.font = UIFont.systemFont(ofSize: 14)
+        placeholderLabel.textColor = .placeholderTextGreyColor
+        placeholderLabel.sizeToFit()
+        placeholderLabel.isHidden = !messageTextView.text.isEmpty
+        
+        NSLayoutConstraint.activate([
+            placeholderLabel.topAnchor.constraint(equalTo: messageTextView.topAnchor, constant: 12),
+            placeholderLabel.leadingAnchor.constraint(equalTo: messageTextView.leadingAnchor, constant: 34),
+            placeholderLabel.bottomAnchor.constraint(equalTo: messageTextView.bottomAnchor, constant: -12),
+        ])
+    }
+    
+    private func setUpSendButton() {
+        addSubview(sendButton)
+        sendButton.setImage(UIImage(named: "sendButton"), for: .normal)
+        sendButton.addTarget(self, action: #selector(sendButtonPressed), for: .touchUpInside)
+        sendButton.setWidth(Constants.sendButtonWidth)
+        sendButton.setHeight(Constants.sendButtonHeight)
+        
+        NSLayoutConstraint.activate([
+            sendButton.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -27),
+            sendButton.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -24)
+        ])
+    }
+    
+    @objc func sendButtonPressed() {
+        // Add your send button action here
+    }
+}
+
+extension TypingAreaView: UITextViewDelegate {
+    
+    func textViewDidChange(_ textView: UITextView) {
+        // Update placeholder label
+        if let placeholderLabel = textView.subviews.first(where: { $0 is UILabel }) as? UILabel {
+            placeholderLabel.isHidden = !textView.text.isEmpty
+        }
+        //update message text view height + scroll
+        guard let lineHeight = textView.font?.lineHeight else { return }
+        let maxHeight: CGFloat = lineHeight * 4 + textView.textContainerInset.top + textView.textContainerInset.bottom
+        let fixedWidth = textView.frame.size.width
+        
+        let newSize = textView.sizeThatFits(CGSize(width: fixedWidth, height: CGFloat.greatestFiniteMagnitude))
+        let clampedHeight = min(newSize.height, maxHeight)
+        
+        var newFrame = textView.frame
+        newFrame.size = CGSize(width: max(newSize.width, fixedWidth), height: clampedHeight)
+        textView.frame = newFrame
+        
+        textView.isScrollEnabled = newSize.height >= maxHeight
```

Comment:
I would consider making two functions updatePlaceholder() and updateTextViewHeight() and would call it here, (naming is up to you).

and what happens when textView gets bigger and then user delete the message? is textView height shrinking or staying the same ? :P

Reply:
yes in this way text view is shrinking but when i split textViewDidChange function, placeholder isn't removing and text view height has problem with number of lines (after 3 its not typing)